Error: Cannot find module 'xxx'

### Issue workflow progress

<!-- PLEASE DO NOT REMOVE THIS SECTION -->

_Progress of the issue based on the
[Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [ ] 1. The issue provides a reproduction available on
     [Github](https://github.com/ardatan/graphql-mesh/tree/master/examples/hello-world),
     [Stackblitz](https://stackblitz.com/github/ardatan/graphql-mesh/tree/master/examples/hello-world)
     or
     [CodeSandbox](https://codesandbox.io/s/github/ardatan/graphql-mesh/tree/master/examples/hello-world)

> Make sure to fork this template and run `yarn generate` in the terminal.
>
> Please make sure Mesh package versions under `package.json` matches yours.

- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

```bash
 Error: Cannot find module '@graphql-mesh/cache-localforage'
      at <unknown> (graphcommerce/examples/magento-graphcms/.next/server/chunks/6807.js:1:118)
      at async T (graphcommerce/examples/magento-graphcms/.next/server/chunks/6807.js:1:1749) {
    code: 'MODULE_NOT_FOUND'
  }
```

It definitely is installed and we don't have any nested node_modules, cleared the yarn.lock and the node_modules directories. It seems the mesh build works but when using it inside next.js it is failing.

If we comment out the cache-localforage in the .mesh.ts it breaks on the next import. We can consistently reproduce it on multiple deveices and multiple node versions.


<!-- A clear and concise description of what the bug is. -->

**To Reproduce** Steps to reproduce the behavior:

Import the .mesh in a nextjs project and try to load a page it will throw the error.

<!-- Adding a codesandbox can help us understand the bug better and speed up things -->

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->

If we replace the occurences of 

```ts
await defaultImportFn(moduleName)`
```

with

```ts
await import(moduleName).then((m) => m.default || m)
```

everything works as expected and it builds. Maybe something changed in the latest release?

**Environment:**

- OS: MacOS
- `@graphql-mesh/...`: latest
- NodeJS: 18 / 20

**Additional context**

If I go to a release before this commit https://github.com/ardatan/graphql-mesh/commit/93fb3643d90c52084725d79b586ace8ecd570911 it starts working again.

<!-- Add any other context about the problem here. -->

--- a/.changeset/plenty-trees-live.md
+++ b/.changeset/plenty-trees-live.md
@@ -4,3 +4,4 @@
 ---
 
 Fix support for bundling environments like Next.js
+Regression introduced in [#8082](https://github.com/ardatan/graphql-mesh/pull/8082)

--- a/.changeset/plenty-trees-live.md
+++ b/.changeset/plenty-trees-live.md
@@ -1,0 +1,6 @@
+---
+'@graphql-mesh/config': patch
+'@graphql-mesh/utils': patch
+---
+
+Fix support for bundling environments like Next.js

--- a/packages/legacy/config/src/process.ts
+++ b/packages/legacy/config/src/process.ts
@@ -124,7 +124,7 @@ export async function processConfig(
   if (config.require) {
     await Promise.all(config.require.map(mod => importFn(mod)));
     for (const mod of config.require) {
-      importCodes.add(`import '${mod}';`);
+      codes.add(`await import('${mod}');`);
     }
   }
 
@@ -208,11 +208,10 @@ export async function processConfig(
             additionalPrefixes: additionalPackagePrefixes,
           }).then(({ resolved: HandlerCtor, moduleName }) => {
             if (options.generateCode) {
-              const handlerImportName = pascalCase(handlerName + '_Handler');
+              const handlerImportName = pascalCase(handlerVariableName);
               codes.add(
-                `const ${handlerImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);`,
-              );
-              codes.add(`const ${handlerVariableName} = new ${handlerImportName}({
+                `const ${handlerImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);\n` +
+                  `const ${handlerVariableName} = new ${handlerImportName}({
               name: ${JSON.stringify(source.name)},
               config: ${JSON.stringify(handlerConfig)},
               baseDir,
@@ -221,7 +220,8 @@ export async function processConfig(
               store: sourcesStore.child(${JSON.stringify(source.name)}),
               logger: logger.child(${JSON.stringify(source.name)}),
               importFn,
-            });`);
+            });`,
+              );
             }
             return new HandlerCtor({
               name: source.name,
@@ -248,11 +248,10 @@ export async function processConfig(
                 });
 
               if (options.generateCode) {
-                const transformImportName = pascalCase(transformName + '_Transform');
+                const transformImportName = pascalCase(transformsVariableName);
                 codes.add(
-                  `const ${transformImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);`,
-                );
-                codes.add(`${transformsVariableName}[${transformIndex}] = new ${transformImportName}({
+                  `const ${transformImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);\n` +
+                    `${transformsVariableName}[${transformIndex}] = new ${transformImportName}({
                   apiName: ${JSON.stringify(source.name)},
                   config: ${JSON.stringify(transformConfig)},
                   baseDir,
@@ -260,7 +259,8 @@ export async function processConfig(
                   pubsub,
                   importFn,
                   logger,
-                });`);
+                });`,
+                );
               }
 
               return new TransformCtor({
@@ -304,10 +304,10 @@ export async function processConfig(
         });
 
         if (options.generateCode) {
-          const transformImportName = pascalCase(transformName + '_Transform');
-          importCodes.add(`import ${transformImportName} from ${JSON.stringify(moduleName)};`);
-
-          codes.add(`transforms[${transformIndex}] = new (${transformImportName} as any)({
+          const transformImportName = pascalCase('Root_Transform_' + transformIndex);
+          codes.add(
+            `const ${transformImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);\n` +
+              `transforms[${transformIndex}] = new ${transformImportName}({
             apiName: '',
             config: ${JSON.stringify(transformConfig)},
             baseDir,
@@ -315,7 +315,8 @@ export async function processConfig(
             pubsub,
             importFn,
             logger,
-          })`);
+          })`,
+          );
         }
         return new TransformLibrary({
           apiName: '',

--- a/packages/legacy/config/src/process.ts
+++ b/packages/legacy/config/src/process.ts
@@ -104,7 +104,7 @@ export async function processConfig(
   const importCodes = new Set([
     `import type { GetMeshOptions } from '@graphql-mesh/runtime';`,
     `import type { YamlConfig } from '@graphql-mesh/types';`,
-    `import { defaultImportFn } from '@graphql-mesh/utils';`,
+    `import { defaultImportFn, handleImport } from '@graphql-mesh/utils';`,
   ]);
   const codes = new Set([
     `export const rawServeConfig: YamlConfig.Config['serve'] = ${JSON.stringify(
@@ -210,7 +210,7 @@ export async function processConfig(
             if (options.generateCode) {
               const handlerImportName = pascalCase(handlerName + '_Handler');
               codes.add(
-                `const ${handlerImportName} = await defaultImportFn(${JSON.stringify(moduleName)});`,
+                `const ${handlerImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);`,
               );
               codes.add(`const ${handlerVariableName} = new ${handlerImportName}({
               name: ${JSON.stringify(source.name)},
@@ -250,7 +250,7 @@ export async function processConfig(
               if (options.generateCode) {
                 const transformImportName = pascalCase(transformName + '_Transform');
                 codes.add(
-                  `const ${transformImportName} = await defaultImportFn(${JSON.stringify(moduleName)});`,
+                  `const ${transformImportName} = await import(${JSON.stringify(moduleName)}).then(handleImport);`,
                 );
                 codes.add(`${transformsVariableName}[${transformIndex}] = new ${transformImportName}({
                   apiName: ${JSON.stringify(source.name)},
@@ -337,7 +337,7 @@ export async function processConfig(
           if (options.generateCode) {
             const importProp = `[${JSON.stringify(importName)}]`;
             codes.add(
-              `const ${importName} = await defaultImportFn(${JSON.stringify(moduleName)}).then(m => m?.${importProp});`,
+              `const ${importName} = await import(${JSON.stringify(moduleName)}).then(m => m?.${importProp});`,
             );
             codes.add(
               `additionalEnvelopPlugins[${pluginIndex}] = await ${importName}(${JSON.stringify(
@@ -367,7 +367,7 @@ export async function processConfig(
           if (options.generateCode) {
             const importName = camelCase('use_' + pluginName);
             codes.add(
-              `const ${importName} = await defaultImportFn(${JSON.stringify(moduleName)});`,
+              `const ${importName} = await import(${JSON.stringify(moduleName)}).then(handleImport);`,
             );
             codes.add(`additionalEnvelopPlugins[${pluginIndex}] = await ${importName}({
           ...(${JSON.stringify(pluginConfig, null, 2)}),
@@ -548,7 +548,9 @@ export async function processConfig(
 
   const mergerLoggerPrefix = `${mergerName}Merger`;
   if (options.generateCode) {
-    codes.add(`const Merger = await defaultImportFn(${JSON.stringify(mergerModuleName)});`);
+    codes.add(
+      `const Merger = await import(${JSON.stringify(mergerModuleName)}).then(handleImport);`,
+    );
     codes.add(`const merger = new Merger({
         cache,
         pubsub,
@@ -566,9 +568,9 @@ export async function processConfig(
 
   if (config.additionalEnvelopPlugins) {
     codes.add(
-      `const importedAdditionalEnvelopPlugins = await defaultImportFn(${JSON.stringify(
+      `const importedAdditionalEnvelopPlugins = await import(${JSON.stringify(
         pathModule.join('..', config.additionalEnvelopPlugins).split('\\').join('/'),
-      )});`,
+      )}).then(handleImport);`,
     );
     const importedAdditionalEnvelopPlugins = await importFn(
       pathModule.isAbsolute(config.additionalEnvelopPlugins)
@@ -649,7 +651,7 @@ export async function processConfig(
         ${[...documentHashMapCodes].join(',\n')}
       }`);
       codes.add(
-        `const usePersistedOperations = await defaultImportFn('@graphql-yoga/plugin-persisted-operations').then(m => m?.usePersistedOperations);`,
+        `const usePersistedOperations = await import('@graphql-yoga/plugin-persisted-operations').then(m => m?.usePersistedOperations);`,
       );
       codes.add(`additionalEnvelopPlugins.push(usePersistedOperations({
         getPersistedOperation(key) {

--- a/packages/legacy/config/src/utils.ts
+++ b/packages/legacy/config/src/utils.ts
@@ -120,7 +120,7 @@ export async function resolveCustomFetch({
   if (!fetchConfig) {
     return {
       fetchFn: defaultFetch,
-      code: `const fetchFn = await defaultImportFn('@whatwg-node/fetch').then(m => m?.fetch || m);`,
+      code: `const fetchFn = await import('@whatwg-node/fetch').then(m => m?.fetch || m);`,
     };
   }
   const { moduleName, resolved: fetchFn } = await getPackage<MeshFetch>({
@@ -132,7 +132,7 @@ export async function resolveCustomFetch({
   });
 
   const processedModuleName = moduleName.startsWith('.') ? path.join('..', moduleName) : moduleName;
-  const code = `const fetchFn = await defaultImportFn(${JSON.stringify(processedModuleName)}).then(m => m?.fetch || m);`;
+  const code = `const fetchFn = await import(${JSON.stringify(processedModuleName)}).then(m => m?.fetch || m);`;
 
   return {
     fetchFn,
@@ -173,7 +173,7 @@ export async function resolveCache(
     logger,
   });
 
-  const code = `const MeshCache = await defaultImportFn(${JSON.stringify(moduleName)});
+  const code = `const MeshCache = await import(${JSON.stringify(moduleName)}).then(handleImport);
   const cache = new MeshCache({
       ...${JSON.stringify(config)},
       importFn,
@@ -218,7 +218,7 @@ export async function resolvePubSub(
 
     const pubsub = new PubSub(pubsubConfig);
 
-    const code = `const PubSub = await defaultImportFn(${JSON.stringify(moduleName)});
+    const code = `const PubSub = await import(${JSON.stringify(moduleName)}).then(handleImport);
     const pubsub = new PubSub(${JSON.stringify(pubsubConfig)});`;
 
     return {
@@ -280,7 +280,7 @@ export async function resolveLogger(
     return {
       logger,
       importCode: ``,
-      code: `const logger = await defaultImportFn(${JSON.stringify(processedModuleName)});`,
+      code: `const logger = await import(${JSON.stringify(processedModuleName)}).then(handleImport);`,
     };
   }
   const logger = new DefaultLogger(initialLoggerPrefix);

--- a/packages/legacy/config/test/processConfig.test.ts
+++ b/packages/legacy/config/test/processConfig.test.ts
@@ -28,7 +28,7 @@ describe('processConfig', () => {
     let codesIteratorResult = codesIterator.next();
     let includesCustomFetch;
     while (!codesIteratorResult.done) {
-      if (codesIteratorResult.value.startsWith('const fetchFn = await defaultImportFn(')) {
+      if (codesIteratorResult.value.startsWith('const fetchFn = await import(')) {
         meshConfigContent = meshConfigContent.concat(codesIteratorResult.value, '\n');
         includesCustomFetch = true;
         break;
@@ -42,7 +42,7 @@ describe('processConfig', () => {
     // Adding export of fetch function so its resolution is actually attempted
     meshConfigContent = meshConfigContent.concat('export { fetchFn };', '\n');
 
-    meshConfigContent = meshConfigContent.replace('await defaultImportFn', 'fakeImport');
+    meshConfigContent = meshConfigContent.replace('await import', 'fakeImport');
     meshConfigContent =
       'const fakeImport = m => require("@graphql-tools/utils").fakePromise(require(m));\n' +
       meshConfigContent;

--- a/packages/legacy/utils/src/defaultImportFn.ts
+++ b/packages/legacy/utils/src/defaultImportFn.ts
@@ -36,13 +36,13 @@ function defaultImportFn(path: string): Promise<any> {
   );
 }
 
-function handleImport(module: any): any {
+function handleImport<T>(module: T): T {
   let i = 0;
-  while (module?.default != null) {
-    if (i > 10 || module === module.default) {
+  while ((module as any)?.default != null) {
+    if (i > 10 || module === (module as any).default) {
       break;
     }
-    module = module.default;
+    module = (module as any).default;
     i++;
   }
   if (typeof module === 'object' && module != null) {
@@ -52,7 +52,7 @@ function handleImport(module: any): any {
       for (const key in module) {
         normalizedVal[key] = module[key];
       }
-      return normalizedVal;
+      return normalizedVal as T;
     }
   }
   return module;

--- a/packages/legacy/utils/src/defaultImportFn.ts
+++ b/packages/legacy/utils/src/defaultImportFn.ts
@@ -36,4 +36,26 @@ function defaultImportFn(path: string): Promise<any> {
   );
 }
 
-export { defaultImportFn };
+function handleImport(module: any): any {
+  let i = 0;
+  while (module?.default != null) {
+    if (i > 10 || module === module.default) {
+      break;
+    }
+    module = module.default;
+    i++;
+  }
+  if (typeof module === 'object' && module != null) {
+    const prototypeOfObject = Object.getPrototypeOf(module);
+    if (prototypeOfObject == null || prototypeOfObject === Object.prototype) {
+      const normalizedVal: Record<string, any> = {};
+      for (const key in module) {
+        normalizedVal[key] = module[key];
+      }
+      return normalizedVal;
+    }
+  }
+  return module;
+}
+
+export { defaultImportFn, handleImport };

